Sprint2/step 1

i0ji opened the pull request:
Intermediate work with Context;

Commit:
31.01.24/b:sprint2/step-1/v:0.1.6.1
unused ERROR in App fixed;
context created/starting to use context

Commit:
31.01.24/b:sprint2/step-1/v:0.1.6.1

Commit:
31.01.24/b:sprint2/step-1/v:0.1.6.1

i0ji committed:
01.02.24/b:sprint2/step-1/v:0.1.6.2:
Data randomizer function added; minor changes to the layout and size of components

i0ji committed:
01.02.24/b:sprint2/step-1/v:0.1.6.2:
One-way context pushed to Ingredients Block

i0ji committed:
02.02.24/b:sprint2/step-1/v:0.1.6.2:
sum function;data request;

i0ji committed:
v:0.1.6.2:
order ID is passed to OrderDetailsModal using Context.

i0ji committed:
v:0.1.6.2:
added Context to App-Constructor coop

gennady-bars commented on src/App.tsx:
```diff
@@ -3,44 +3,82 @@ import appStyles from './App.module.scss';
+			<main className={appStyles.burger_builder}>
+				{state.error ? (<p>Произошла ошибка: {error}</p>) : (ingredientsData.length > 0 && (
+					<>
+						<IngredientsContext.Provider value={state.ingredientsData}>
```
Хорошие общие контексты сделали

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -2,117 +2,146 @@ import burgerConstructorStyles from "./BurgerConstructorStyles.module.scss";
-            {isVisible &&
+export default function BurgerConstructor() {
+	
+	const [isVisible, toggleVisibility] = useReducer(isVisible => !isVisible, false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars commented on src/utils/order-api.ts:
```diff
@@ -0,0 +1,23 @@
+		},
+		body: JSON.stringify(requestBody),
+	})
+		.then(response => {
+			if (!response.ok) {
+				throw new Error('Something went wrong');
+			}
+			return response.json();
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужного повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/utils/order-api.ts:
```diff
@@ -0,0 +1,23 @@
+export const orderUrl = 'https://norma.nomoreparties.space/api/orders'
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -2,117 +2,146 @@ import burgerConstructorStyles from "./BurgerConstructorStyles.module.scss";
+					console.error('YOU WILL NOT GET FOOD:', responseData);
+				}
+			})
+			.catch(error => {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -2,117 +2,146 @@ import burgerConstructorStyles from "./BurgerConstructorStyles.module.scss";
                 <>
                     <Modal onClose={handleCloseModal}>
-                        <OrderDetails/>
+                        <OrderDetailsContext.Provider value={orderNumberContext!}>
```
`OrderDetailsContext` не имеет смысла


Контекст нужен тогда, когда его значение используется как минимум в 2х и более компонентах (и / или компонент, которому нужно это значение, находится очень глубоко по вложенности от главного компонента, где появляются эти данные). Если же он только в одном компоненте используется и недалеко находится компонент, которому нужны данные эти, тогда нет особого смысла делать контекст.

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -2,117 +2,146 @@ import burgerConstructorStyles from "./BurgerConstructorStyles.module.scss";
+			{/* ----- MODAL ENTER ----- */}
+			
+			{isVisible &&
                 <>
```
`<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Хорошие общие контексты сделали
* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу


но есть некоторые недочеты:




* У булок не бывает `DragIcon` по макету  https://disk.yandex.ru/i/X9T_7kIJfX5Zpg 
* Верхняя и нижняя булки должны быть всегда **одинаковыми** по заданию и у них должна быть иконка `Замочек`, так как их нельзя будет удалять (только заменять перетаскиванием)
* Буква `Я` в заказе https://disk.yandex.ru/i/6b7KPTlQxZ5aMw 
* Внутри `AppHeader` должен быть тег `header` вместо `div` по семантике
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* Контекст нужен тогда, когда его значение используется как минимум в 2х и более компонентах (и / или компонент, которому нужно это значение, находится очень глубоко по вложенности от главного компонента, где появляются эти данные). Если же он только в одном компоненте используется и недалеко находится компонент, которому нужны данные эти, тогда нет особого смысла делать контекст.
* `<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку
* Для атрибута `key` лучше всего использовать уникальный `id` : у карточки как раз есть такой только с нижним подчеркиванием `_id`  [Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)  


**Можно лучше**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента

i0ji committed:
v:0.1.6.2:
added Context to App-Constructor coop

Commit:
v:0.1.6.3
API request separated into a separated file in src/utils; added error catching state;
refactoring of ingredient types filtering in BurgerIngs component using useMemo hook;

Commit:
v:0.1.6.3
multiple code edits;
checkResponse func added; added custom hook to manage orderModal visibility and order ID request;

Commit:
v:0.1.6.3
multiple code edits;
README.md;
